File: packages/schemas/src/task-notifications.ts
import z from "zod/v4";

export const TASK_NOTIFICATION_PROTOCOL_VERSION = 1;

export const taskNotificationCategorySchema = z.enum([
  "decisions",
  "failures",
  "completions",
  "assignments",
]);

export type TaskNotificationCategory = z.infer<typeof taskNotificationCategorySchema>;

export const taskNotificationPreferencesSchema = z.object({
  enabled: z.boolean(),
  decisions: z.boolean(),
  failures: z.boolean(),
  completions: z.boolean(),
  assignments: z.boolean(),
});

export type TaskNotificationPreferences = z.infer<typeof taskNotificationPreferencesSchema>;

export const updateTaskNotificationPreferencesSchema = taskNotificationPreferencesSchema.partial();

export type UpdateTaskNotificationPreferences = z.infer<
  typeof updateTaskNotificationPreferencesSchema
>;

export const notificationPermissionSchema = z.enum(["prompt", "granted", "denied", "unsupported"]);

export type NotificationPermission = z.infer<typeof notificationPermissionSchema>;

export const taskNotificationRegistrationStateSchema = z.enum(["registered", "failed", "disabled"]);

export const taskNotificationPlatformSchema = z.literal("web");
export type TaskNotificationPlatform = z.infer<typeof taskNotificationPlatformSchema>;

const registrationFields = {
  installationId: z.string().trim().min(1).max(128),
};

export const registerTaskNotificationSchema = z.object({
  ...registrationFields,
  platform: z.literal("web"),
  subscription: z.object({
    endpoint: z.url().max(2048),
    expirationTime: z.number().nonnegative().nullable().default(null),
    keys: z.object({
      p256dh: z.string().trim().min(1).max(1024),
      auth: z.string().trim().min(1).max(1024),
    }),
  }),
});

export type RegisterTaskNotification = z.infer<typeof registerTaskNotificationSchema>;

export const taskNotificationRegistrationSchema = z.object({
  id: z.string().min(1),
  installationId: z.string().min(1),
  platform: taskNotificationPlatformSchema,
  state: taskNotificationRegistrationStateSchema,
  failureCode: z.string().nullable(),
  updatedAt: z.string(),
});

export type TaskNotificationRegistration = z.infer<typeof taskNotificationRegistrationSchema>;

export const taskNotificationSettingsSchema = z.object({
  protocolVersion: z.literal(TASK_NOTIFICATION_PROTOCOL_VERSION),
  preferences: taskNotificationPreferencesSchema,
  registrations: z.array(taskNotificationRegistrationSchema),
  webPushPublicKey: z.string().nullable(),
});

export type TaskNotificationSettings = z.infer<typeof taskNotificationSettingsSchema>;

export const taskNotificationRegistrationResponseSchema = z.object({
  registration: taskNotificationRegistrationSchema,
});

export const taskInboxReceiptInputSchema = z.object({
  itemIds: z.array(z.string().min(1).max(256)).min(1).max(100),
});

export const taskInboxMutationResponseSchema = z.object({
  updated: z.number().int().nonnegative(),
});

export const taskNotificationDeepLinkSchema = z.object({
  protocolVersion: z.literal(TASK_NOTIFICATION_PROTOCOL_VERSION),
  itemId: z.string().min(1),
  current: z.boolean(),
  deepLink: z.string().nullable(),
});

export type TaskNotificationDeepLink = z.infer<typeof taskNotificationDeepLinkSchema>;

export function createTaskInboxItemId(taskId: string, taskVersion: number): string {
  return `${taskId}:v${taskVersion}`;
}

export function parseTaskInboxItemId(
  itemId: string,
): { taskId: string; taskVersion: number } | null {
  const match = /^(.+):v([1-9]\d*)$/.exec(itemId);

  if (!match) {
    return null;
  }

  const taskVersion = Number(match[2]);

  return Number.isSafeInteger(taskVersion) ? { taskId: match[1], taskVersion } : null;
}

File: apps/mobile/ios/Polychat/Models/TaskNotifications.swift
import Foundation

struct TaskInboxItem: Codable, Equatable, Identifiable {
    let id: String
    let kind: String
    let taskId: String
    let projectId: String
    let workspaceId: String
    let projectName: String
    let objective: String
    let detail: String?
    let conversationId: String?
    let since: String
    let requiresAction: Bool
    let isRead: Bool
    let readAt: String?
    let deepLink: String
}

struct TaskInboxResponse: Codable, Equatable {
    let items: [TaskInboxItem]
    let total: Int
    let unread: Int
}

struct TaskNotificationPreferences: Codable, Equatable {
    var enabled: Bool
    var decisions: Bool
    var failures: Bool
    var completions: Bool
    var assignments: Bool
}

struct TaskNotificationRegistration: Codable, Equatable, Identifiable {
    let id: String
    let installationId: String
    let platform: String
    let state: String
    let failureCode: String?
    let updatedAt: String
}

struct TaskNotificationSettings: Codable, Equatable {
    let protocolVersion: Int
    var preferences: TaskNotificationPreferences
    let registrations: [TaskNotificationRegistration]
    let webPushPublicKey: String?
}

struct UpdateTaskNotificationPreferencesRequest: Encodable {
    let enabled: Bool?
    let decisions: Bool?
    let failures: Bool?
    let completions: Bool?
    let assignments: Bool?

    init(
        enabled: Bool? = nil,
        decisions: Bool? = nil,
        failures: Bool? = nil,
        completions: Bool? = nil,
        assignments: Bool? = nil
    ) {
        self.enabled = enabled
        self.decisions = decisions
        self.failures = failures
        self.completions = completions
        self.assignments = assignments
    }
}

struct TaskInboxReceiptRequest: Encodable {
    let itemIds: [String]
}

struct TaskInboxMutationResponse: Decodable {
    let updated: Int
}
